src/platform/silabs/efr32/KeyValueStoreManagerImpl.cpp

```diff
@@ -67,25 +70,63 @@ bool KeyValueStoreManagerImpl::IsValidKvsNvm3Key(uint32_t nvm3Key) const
     return ((SilabsConfig::kConfigKey_KvsFirstKeySlot <= nvm3Key) && (nvm3Key <= SilabsConfig::kConfigKey_KvsLastKeySlot));
 }
 
-CHIP_ERROR KeyValueStoreManagerImpl::MapKvsKeyToNvm3(const char * key, uint32_t & nvm3Key, bool isSlotNeeded) const
+uint16_t KeyValueStoreManagerImpl::hashKvsKeyString(const char * key) const
+{
+    uint8_t hash256[Crypto::kSHA256_Hash_Length] = { 0 };
+    Crypto::Hash_SHA256(reinterpret_cast<const uint8_t *>(key), strlen(key), hash256);
+
+    uint16_t hash16, i = 0;
+
+    while (!hash16 && (i < (Crypto::kSHA256_Hash_Length - 1)))
+    {
+        hash16 = (hash256[i] | (hash256[i + 1] << 8));
+        i++;
+    }
+    return hash16;
+}
+
+CHIP_ERROR KeyValueStoreManagerImpl::MapKvsKeyToNvm3(const char * key, uint16_t hash, uint32_t & nvm3Key, bool isSlotNeeded) const
 {
     CHIP_ERROR err;
+    char * strPrefix          = nullptr;
     uint8_t firstEmptyKeySlot = kMaxEntries;
     for (uint8_t keyIndex = 0; keyIndex < kMaxEntries; keyIndex++)
     {
-        if (strcmp(key, mKvsStoredKeyString[keyIndex]) == 0)
+        if (mKvsKeyMap[keyIndex] == hash)
         {
-            nvm3Key = CONVERT_KEYMAP_INDEX_TO_NVM3KEY(keyIndex);
-            VerifyOrDie(IsValidKvsNvm3Key(nvm3Key) == true);
-            return CHIP_NO_ERROR;
+            uint32_t tempNvm3key = CONVERT_KEYMAP_INDEX_TO_NVM3KEY(keyIndex);
+            VerifyOrDie(IsValidKvsNvm3Key(tempNvm3key) == true);
+
+            size_t readCount;
+            size_t length = strlen(key);
+            if (strPrefix == nullptr)
```

Comment:
Could that check be move out of the `for` loop to be more efficient ? Since we either complete the loop or return after freeing the buffer the allocation could be done once outside and save some processing.

Reply:
I did it this way so I only allocate if the hash is found/key was stored present. If the hash is never found we don't need to allocate. Also in the higlhly rare case where a collision would occur, the alloc would already be done this way so we don't have to do it again. 

To make it short. The check and alloc will be done onece 99% of the time. Else it happens twice.

Reply:
Thanks for the answer